src/operations/query.ts

```diff
@@ -71,9 +73,104 @@ export const startQuery = (store: Store, request: OperationRequest) => {
   };
 
   result.data = readSelection(ctx, 'Query', getSelectionSet(operation), root);
+
+  return result;
+};
+
+export const readOperation = (
+  store: Store,
+  request: OperationRequest,
+  data: Data
+) => {
+  initStoreState(0);
+  const operation = getMainOperation(request.query);
+  const root: Data = Object.create(null);
+  const result: QueryResult = {
+    completeness: 'FULL',
+    dependencies: getCurrentDependencies(),
+    data: root,
+  };
+
+  const operationName = getOperationName(operation);
+  const ctx: Context = {
+    variables: normalizeVariables(operation, request.variables),
+    fragments: getFragments(request.query),
+    result,
+    store,
+  };
+
+  root.__typename = operationName;
+  result.data = readRoot(
+    ctx,
+    operationName,
+    getSelectionSet(operation),
+    result.data as Data,
+    data
+  );
+  clearStoreState();
   return result;
 };
 
+export const readRoot = (
+  ctx: Context,
+  entityKey: string,
+  select: SelectionSet,
+  data: Data,
+  originalData: Data | NullArray<Data>
+): Data => {
+  const iter = new SelectionIterator(entityKey, entityKey, select, ctx);
+  let node;
+  while ((node = iter.next()) !== undefined) {
+    const fieldAlias = getFieldAlias(node);
+    const fieldValue = originalData[fieldAlias];
+
+    if (
+      node.selectionSet !== undefined &&
+      fieldValue !== null &&
+      !isScalar(fieldValue)
+    ) {
+      data[fieldAlias] = Array.isArray(fieldValue)
```

Comment:
If I'm not mistaken, since the original data is passed in separately, this statement of copying isn't needed anymore, right?